File: LeetCode/binary_tree_max_path_sum.rb
#  Given a binary tree, find the maximum path sum.
#
#  The path may start and end at any node in the tree. 
#

class Node < Struct.new(:value, :left, :right)
end

root = Node.new(1,
                Node.new(2, nil, nil),
                Node.new(3, nil, nil))

def max_sum(root, max) 
    return 0 if root.nil?
    left = max_sum(root.left, max)
    right = max_sum(root.right, max)
    max_sum = [root.value, left, left + root.value, right + root.value, left + right + root.value].max
    max_top = [root.value, left + root.value, right + root.value].max
    max[0] = max_sum
    return max_top
end

max = []
max_sum(root, max)
puts max.inspect

File: LeetCode/remote/binary_tree_max_path_sum.cs
//  https://leetcode.com/problems/binary-tree-maximum-path-sum/
//
//   Given a binary tree, find the maximum path sum.
//
//   For this problem, a path is defined as any sequence of nodes
//   from some starting node to any node in the tree along the parent-child connections.
//   The path does not need to go through the root.

using System;

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution {
    //  
    //  Submission Details
    //  92 / 92 test cases passed.
    //      Status: Accepted
    //      Runtime: 204 ms
    //          
    //          Submitted: 0 minutes ago
    //
    public int MaxPathSum(TreeNode root) {
        var result = int.MinValue;
        Max(root, ref result);
        return result;
    }

    public int Max(TreeNode root, ref int result)
    {
        var left = root.left == null ? 0 : Max(root.left, ref result);
        var right = root.right == null ? 0 : Max(root.right, ref result);
        left = left < 0 ? 0 : left;
        right = right < 0 ? 0 : right;
        result = Math.Max(result, left + right + root.val);
        return root.val + Math.Max(left, right);
    }

    static void Main()
    {
        Console.WriteLine(new Solution().MaxPathSum(new TreeNode(1) {
                        left = new TreeNode(2),
                        right = new TreeNode(3)
                    }));
        Console.WriteLine(new Solution().MaxPathSum(new TreeNode(2) {
                        left = new TreeNode(-1)
                    }));
    }
}
